### DISCUSSION TITLE
How to read temperature from Sonoff SC?

### DISCUSSION THREAD

**Discussion by 00shoham** [arendst/Tasmota#12176] [Q&A] [answered]
I know this is probably a N00b question but I can't for the life of me figure out how to do it.  :-)

I've got a Sonoff SC.  I've flashed it with tasmota-sensors.bin.gz, I can connect to it with a web browser and I can issue commands via the web console.  For example, if I type 'STATUS' I get this JSON response:

22:06:39.153 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_FD1C33","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}

I've selected the Sonoff SC module and template, though to be honest I'm not sure I understand what 'modules' and 'templates' really mean.

What I can't seem to do is figure out how to actually read the state of the sensors on this device - i.e., temperature, pressure, etc.  That's kind of the point of the device and I was hoping to be able to pull that info from the device via web services - but heck, even via a web UI would be fine.  I understand that it's really a secondary device onboard, connected via a serial port, that has the sensors.  Do I have to tell the main device to send commands to the child device and read output back?  If so, how is that done?

Any suggestions?  What silly thing am I missing?  :-)

Thanks!

--
Idan

**Comment by sfromis** [comment, marked as answer]
The template is for easy copy/paste, instead of messing around with configuring both module and pins. Those two things are included in the template, but if you then use Configure Module, you may be missing out on having the pins configured correctly. I'd recommend using the template, and ensure that the Module selection is number 0 which will activate the template. The template by itself then selects Module 21 internally, to activate the part of Tasmota supporting the device.

I have absolutely no experience with Sonoff SC, but my expectation would be that you should get data every TelePeriod. Tasmota is designed for MQTT, which will allow Tasmota to automatically publish the data in a MQTT payload. If you "insist" on the worse method of not using MQTT, the `Status 10` command should also provide the sensor JSON data.

**Reply by 00shoham** [reply]
Thank you very much for the prompt response.

Never mind MQTT versus other methods - once I can see the temperature, I'll work out the best software path to move it out.  :-)

I tried 'STATUS 10' and got back just the time:

00:23:38.587 CMD: Group 0, Index 1, Command "STATUS", Data "10"
00:23:38.595 MQT: stat/Sensor/STATUS10 = {"StatusSNS":{"Time":"2021-05-24T00:23:38"}}

As you can see I've turned up debug levels and even enabled syslogd to see what's going on...

I notice that the device periodically sends back this:

00:25:29.527 SER: Received AT+STATUS?
00:25:39.528 SER: Received AT+STATUS?
00:25:49.528 SER: Received AT+STATUS?
00:25:59.531 SER: Received AT+STATUS?

Seems like communication between the two chips is happening but what the ATMega328P is sending to the ESP8266 is kind of junk (some sort of byte marker).

I've checked the jumpers on the path between the two chips and both the 'TX' and 'RX' ones are shorted across.

I'm starting to wonder if the issue is hardware...  ?

Hmm.

**Reply by 00shoham** [reply]
Hmm.  If both the TX and RX jumpers are in, then the device does not boot up properly.

If TX is on and RX is off, then the device boots, presents the nice Tasmota web UI, but the serial port shows this:

01:10:49.498 SER: Received AT+STATUS?
01:10:49.501 SER: Transmit AT+STATUS=4
01:10:59.500 SER: Received AT+STATUS?
01:10:59.502 SER: Transmit AT+STATUS=4
01:11:09.500 SER: Received AT+STATUS?
01:11:09.503 SER: Transmit AT+STATUS=4
01:11:19.501 SER: Received AT+STATUS?
01:11:19.503 SER: Transmit AT+STATUS=4
01:11:29.501 SER: Received AT+STATUS?
01:11:29.504 SER: Transmit AT+STATUS=4
01:11:39.501 SER: Received AT+STATUS?
01:11:39.504 SER: Transmit AT+STATUS=4

Fun fun.

**Reply by barbudor** [reply]
The serial port is used for the communication between the ESP8266 and the Atmega.
Apparently, the original firmware on the ESP was the standard "wifi modem" from Espressif where the Atmega is the system controller and the ESP is providing wifi features.

In such configuration, serial logs should be disabled with `SerialLog 0` and Tasmota appropriate module should be enabled to take care of the communication between both chips.

From your log, serial log is still enabled and is probably confusing the Atmega firmware.

**Reply by 00shoham** [reply]
Thanks for that.  Unfortunately, I checked and serial logging was not enabled, so that wasn't it.

![image](https://user-images.githubusercontent.com/64545907/119383238-51d9dc00-bc80-11eb-8c99-a18e59b72ca8.png)

Any other ideas?

**Reply by 00shoham** [reply]
I did increase the polling frequency a bit to reduce waiting on the debug process..

![image](https://user-images.githubusercontent.com/64545907/119383516-a8dfb100-bc80-11eb-9108-c4104c672cff.png)

**Reply by sfromis** [reply]
To me, those symptoms might be a sign of missing config details. You did not react to me posting some pointers there. Just to make sure of your config.... What's the output of `Backlog Module; Template` ?

**Reply by 00shoham** [reply]
Thanks again.  Lets see if I get that right:

Here are the configured modules, via the web UI:

![image](https://user-images.githubusercontent.com/64545907/119403699-fb7a9680-bc9b-11eb-94bc-8cb8938f00b6.png)

Here is the template config, again from the web UI:

![image](https://user-images.githubusercontent.com/64545907/119403783-1d741900-bc9c-11eb-9abc-638a174197e1.png)

When I type "Backlog Module" from the web console it shows:

21:27:20.504 CMD: Group 0, Index 1, Command "BACKLOG", Data "Module"
21:27:20.550 SRC: Backlog
21:27:20.553 CMD: Group 0, Index 1, Command "MODULE", Data ""
21:27:20.560 MQT: stat/Sensor/RESULT = {"Module":{"0":"Sonoff SC"}}

When I type "Template" I get:

21:28:13.778 CMD: Group 0, Index 1, Command "TEMPLATE", Data ""
21:28:13.784 MQT: stat/Sensor/RESULT = {"NAME":"Sonoff SC","GPIO":[32,3200,1,3232,0,0,0,0,0,320,0,0,0,0],"FLAG":0,"BASE":21}

I don't know enough to say whether that's right or wrong..?

**Reply by sfromis** [reply]
Appears to be correct, matching the [templates repository](https://templates.blakadder.com/sonoff_SC.html) setup.

The [driver](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_04_snfsc.ino) comments seem to indicate that `AT+STATUS?` should have Tasmota respond with `AT+STATUS=4` which is all your logs shows of activity. Maybe iTead changed the protocol requiring an updated driver, but that's just pure speculation from my side.

The first serial communication right after restart would probably show a number of other interactions for setting up. Maybe clues there....

**Reply by 00shoham** [reply]
Thanks for sticking with me.  I restarted but didn't see anything terribly interesting..

![image](https://user-images.githubusercontent.com/64545907/119421857-e6622f80-bcbc-11eb-82b5-028359cca386.png)

**Reply by 00shoham** [reply]
You may be right - the serial communication protocol may have changed, I've no idea what it's supposed to be, but it's certainly not getting temperature data back (or other sensor data).

**Reply by barbudor** [reply]
If you can't find a way to get it work with Itead original firmware, it should be possible to reflash the Atmega with an Arduino sketch which handles the sensors and use Tasmota Client. protocol.

**Reply by sfromis** [reply]
Seeing the absence of other serial communication after restart, I'm not convinced that Tasmota works as expected. Of course, having no experience with the device, this is by no means an assertion. As it can happen that refactoring for code maintenance reasons introduces a bug, this is one of the few situations where I'd suggest checking with older Tasmota firmware versions.

**Reply by barbudor** [reply]
This guy wrote an alternative firmware for the ATmega and the repo including a description of the protocol
https://github.com/FastyBird/sonoff-sc

**Comment by arendst** [comment]
Just loaded the latest Tasmota (v9.4.0.4) on my SC and it still works as before. See logging with ``weblog 4`` and a restart:

```
00:00:00.001 HDW: ESP8266EX
00:00:00.046 CFG: Loaded from flash at FB, Count 55
00:00:00.140 CFG: Saved to flash at FB, Count 56, Bytes 4096
00:00:00.141 SER: Set to 8N1 19200 bit/s
00:00:00.206 QPC: Reset
00:00:00.210 SER: Set to 8N1 19200 bit/s
00:00:00.213 SRC: Restart
00:00:00.214 Project tasmota Sc Version 9.4.0.4(theo)-2_7_4_9(2021-05-30T12:39:40)
00:00:00.215 SER: Transmit AT+START
00:00:00.311 CFG: Saved to flash at FA, Count 57, Bytes 4096
00:00:00.468 WIF: Checking connection...
00:00:00.469 WIF: Attempting connection...
00:00:01.262 WIF: Network (re)scan started...
00:00:01.262 WIF: Checking connection...
00:00:01.263 WIF: Attempting connection...
00:00:02.066 SER: Received AT+UPDATE="humidity":88,"temperature":23,"light":9,"noise":2,"dusty":1
00:00:02.067 SER: Transmit AT+SEND=ok
00:00:02.266 WIF: Checking connection...
00:00:02.267 WIF: Attempting connection...
00:00:03.265 WIF: Checking connection...
00:00:03.266 WIF: Attempting connection...
00:00:04.270 WIF: Network 0, AP1, SSId indebuurt_IoT, Channel 1, BSSId 74:83:C2:2A:F1:AC, RSSI -81, Encryption 1
00:00:04.271 WIF: Network 1, AP1, SSId indebuurt_IoT, Channel 11, BSSId 18:E8:29:CA:17:C1, RSSI -4, Encryption 1
00:00:04.609 WIF: Connecting to AP1 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode 11n as sc...
00:00:04.610 WIF: Checking connection...
00:00:04.611 WIF: Attempting connection...
00:00:05.754 WIF: Checking connection...
00:00:05.756 WIF: Connected
00:00:06.005 HTP: Web server active on sc with IP address 192.168.2.143
00:00:06.721 RTC: UTC 2021-05-30T10:41:25, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
12:41:26.025 MQT: Attempting connection...
12:41:26.037 MQT: Connected
12:41:26.040 MQT: tele/sc/LWT = Online (retained)
12:41:26.042 MQT: cmnd/sc/POWER = 
12:41:26.044 MQT: Subscribe to cmnd/sc/#
12:41:26.046 MQT: Subscribe to cmnd/tasmotas/#
12:41:26.047 MQT: Subscribe to cmnd/DVES_F8B22E_fb/#
12:41:26.051 MQT: tele/sc/INFO1 = {"Info1":{"Module":"Sonoff SC","Version":"9.4.0.4(theo)","FallbackTopic":"cmnd/DVES_F8B22E_fb/","GroupTopic":"cmnd/tasmotas/"}}
12:41:26.053 MQT: tele/sc/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"sc","IPAddress":"192.168.2.143"}}
12:41:26.056 MQT: tele/sc/INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
12:41:27.127 SRC: MQTT
12:41:27.129 CMD: Grp 0, Cmnd 'STATUS', Idx 1, Len 2, Data '10'
12:41:27.136 MQT: stat/sc/STATUS10 = {"StatusSNS":{"Time":"2021-05-30T12:41:27","SonoffSC":{"Temperature":23.0,"Humidity":88.0,"DewPoint":20.9,"Light":20,"Noise":20,"AirQuality":100},"TempUnit":"C"}}
12:41:27.602 APP: Boot Count 19
12:41:28.598 CFG: Saved to flash at F9, Count 58, Bytes 4096
12:41:30.483 SER: Received AT+STATUS?
12:41:30.485 SER: Transmit AT+STATUS=4
12:41:30.589 MQT: tele/sc/STATE = {"Time":"2021-05-30T12:41:30","Uptime":"0T00:00:13","UptimeSec":13,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"RSSI":100,"Signal":-6,"LinkCount":1,"Downtime":"0T00:00:07"}}
12:41:30.597 MQT: tele/sc/SENSOR = {"Time":"2021-05-30T12:41:30","SonoffSC":{"Temperature":23.0,"Humidity":88.0,"DewPoint":20.9,"Light":20,"Noise":20,"AirQuality":100},"TempUnit":"C"}
12:41:35.579 MQT: tasmota/discovery/5CCF7FF8B22E/config = {"ip":"192.168.2.143","dn":"Sc","fn":["Sc",null,null,null,null,null,null,null],"hn":"sc","mac":"5CCF7FF8B22E","md":"Sonoff SC","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"9.4.0.4","t":"sc","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":1,"lt_st":0,"sho":[0,0,0,0],"ver":1} (retained)
12:41:35.586 MQT: tasmota/discovery/5CCF7FF8B22E/sensors = {"sn":{"Time":"2021-05-30T12:41:35","SonoffSC":{"Temperature":23.0,"Humidity":88.0,"DewPoint":20.9,"Light":20,"Noise":20,"AirQuality":100},"TempUnit":"C"},"ver":1} (retained)
12:41:40.483 SER: Received AT+STATUS?
12:41:40.486 SER: Transmit AT+STATUS=4
12:41:44.778 WIF: Checking connection...
12:41:47.537 SER: Received AT+UPDATE="humidity":88,"temperature":23,"light":9,"noise":4,"dusty":1
12:41:47.540 SER: Transmit AT+SEND=ok
12:41:50.573 MQT: tele/sc/STATE = {"Time":"2021-05-30T12:41:50","Uptime":"0T00:00:33","UptimeSec":33,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"RSSI":100,"Signal":-8,"LinkCount":1,"Downtime":"0T00:00:07"}}
12:41:50.583 MQT: tele/sc/SENSOR = {"Time":"2021-05-30T12:41:50","SonoffSC":{"Temperature":23.0,"Humidity":88.0,"DewPoint":20.9,"Light":20,"Noise":40,"AirQuality":100},"TempUnit":"C"}
12:41:52.513 SER: Received AT+UPDATE="humidity":88,"temperature":23,"light":9,"noise":2,"dusty":1
12:41:52.517 SER: Transmit AT+SEND=ok
```

**Reply by 00shoham** [reply]
Thank you Theo.

I have to conclude that my problem is hardware, not software.  :-(

Oh well - it was a valiant effort!

**Reply by barbudor** [reply]
It may be possible that this is a more recent Sonoff SC which is not expecting `AT+STATUS=4` as an answer.
Apparently it is not using standard Espressif AT code so probably one of Itead's own.

Did you kept the original firmware ? 
It could be worth to restore it and spy the protocol between the AtMega and the ESP to see if it still matches what Tasmota is expected.

**Reply by 00shoham** [reply]
Alas, I didn't keep the original firmware.  They only sold the SC for a short while - so I don't think there really is a "more recent Sonoff SC."  I had purchased it over a year ago (perhaps 2 years) and got busy with other things, so it just sat there waiting for me to have time to poke at it.  Oh well - it was only like $15.
